### ISSUE TITLE
 Problem with .ts files  when audio changes from dd5.1 to aac or the other way around

### ISSUE THREAD

**Issue by mrmojo666** [xbmc/xbmc#15999]
### Describe the bug
Here is a clear and concise description of what the problem is:

each time i change channel from a dd5.1 to aac2.0 or the other way around, i get crazy buffering and lost audio

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

change channel without issues

## Actual Behavior

each time i change channel from a dd5.1 to aac2.0 or the other way around, i get crazy buffering and lost audio, a channel change to another channel that have the same audio format as the last fixes the problem until again i fall on another audio format change.
change channels between same audio or dd to mpg and viceversa working good
I'm on libreelec raspberry pi 3 b+ 9.0.1, tried also on windows and android TV, always the same beahavior

## Possible Fix

a channel change to another channel that have the same audio format as the last fixes the problem until again i fall on another audio format change
another workround to temporary fix is to toggle on/off passtrough, i think it resets the audio .

### To Reproduce

1. Use mediaportal PVR addon
2. start TV
3. change channel from a Dolby 5.1 channel to aac 2.0 channel (or the other way round)

### Debuglog

The debuglog can be found here:

log part 1 https://paste.kodi.tv/ilogebetuc.kodi
log part 2 https://paste.kodi.tv/okokakedov.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

forum posts:

https://forum.kodi.tv/showthread.php?tid=342732

https://forum.kodi.tv/showthread.php?tid=343177

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [x] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: libreelec 9.0.1
 - Kodi version: 18.1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by Leatherface75** [comment]
This must be a problem with Mediaportal addon because i don't have that problem with TVheadend addon and not seen it on other backends i tried either.
This with Linux, Windows, Android boxes.

**Comment by mrmojo666** [comment]
any clues ? thank you

**Comment by mrmojo666** [comment]
I've spent some more time to try to understand the issue and now i'm sure that is a bug with .ts files.
If I play from video section a .ts where the audio stream changes from aac to ac3 (that is a standard for ts files) kodi stops playing  the file . That is not releted with PVR funcionalities. The same .ts file playing perfectly with VLC, mpchc and other players on any SO. so IMHO it is definitely a KODI videoplayer BUG.

**Comment by mrmojo666** [comment]
@DaVukovic could you please change  tags , removing pvr and add video player?

**Comment by mrmojo666** [comment]
here the log playing a ts fle from video section
https://paste.ubuntu.com/p/9g7W2Hs4cd/

at 30 sec, the video changes the audio format and problem occours

here the ts file https://drive.google.com/open?id=1MGxe7L...ZyK90xbI7d

**Comment by fritsch** [comment]
I checked it with kodi and mpv. On switch theere is pure garbage in the ts file. mpv stops decoding audio entirely and just continues with video. Kodi stalls the moment audio cannot be decoded.

While it is not nice to have a player hang here, please fix the input - you won't have any luck with such garbage sent through the wire.

**Comment by mrmojo666** [comment]
I agree with you there is pure garbage when a channel change occours in that file. I recorded that channel change just to have something that can be tried on other players (with no pvr functions), so i tried on vlc android and windows and mpchc with lav filter and the garbage doesn't breaks the playing. In the normal pvr channel change kodi doesn't stop the video but lost audio (that can be get back just switching on/off audio passthrough). What I start to think is that, maybe the video player of kodi is born as file player so is more sensible to errors in streams than player like vlc that born as network player? I would like to provide you logs with a real pvr channel change (with audio format change, because if the audio doesn't change there are no issues), but i can't give you a file. Or would be useful again a rough recorded ts file with a channel  change with no audio format change?

**Comment by fritsch** [comment]
Nope. Player is made in a way to be adjustable to litterary everything. It's one of the most compatible, extendable architecture / interfaces you can find.

But - there is always a new "king of garbage" which needs a certain adjustment :-) - Your file won.

**Comment by mrmojo666** [comment]
https://drive.google.com/open?id=1FxRhmg-WEKqJtgH8Wxwa8Vif-7nsTCsg

if you want to try more garbage :) , this ts have two channel changes with channel with the same audio format and it does not break kodi  

thank you for the effort

**Comment by FernetMenta** [comment]
The ts file most likely lacks pat/pmt signalling a program change. Without this info videoplayer won't get the trigger to reopen elementary streams belonging to the program. In general different encoding, aac/ac3, don't get mixed on the same audio stream. So if the trigger does not occur, vp stays the the originally selected audio stream.

**Comment by mrmojo666** [comment]
I understand, just wondering how vlc or other directshow based players manage this condition without issues.

**Comment by FernetMenta** [comment]
just noticed that dynamic program switching is broken. not sure if this is the same issue than yours. @mrmojo666 the link the the ts file does not work anymore. could you provide the file again?

**Comment by mrmojo666** [comment]
sure :)  @FernetMenta  https://drive.google.com/file/d/1MGxe7LqNH6_dDobLWiGpHsZyK90xbI7d/view?usp=sharing 

thank you

**Comment by FernetMenta** [comment]
This file is clearly out of spec, as I assumed before. Although ffmpeg (the library Kodi uses for demuxing and decoding) logs tons of errors, it doesn't signal any error to the application level (Kodi). The only error Kodi sees in this case is audio stalling. Kodi tries a seek to re-sync but this file isn't even seekable using timestamps (I guess that the timestamps are not monotonic increasing), so playback jumps to the beginning.

> What I start to think is that, maybe the video player of kodi is born as file player so is more sensible to errors in streams than player like vlc that born as network player?

This applies to ffmpeg. In particular the mpeg-ts demuxer is not the best when it comes to streaming. In this case here a simple internal reset after the 100th error would most likely solve the issue but the demuxer doesn't try any error corrections. Nevertheless, ffmpeg implements the specifications and if some source does not play with ffmpeg, it's most likely a problem with the source.

> I understand, just wondering how vlc or other directshow based players manage this condition without issues.

Those directshow players do not use ffmpeg (vlc doesn't either). Windows world isn't famous for complying to specifications.
Best solution for your problem is to file an issue at mediaportal

**Comment by mrmojo666** [comment]
> This file is clearly out of spec, as I assumed before. Although ffmpeg (the library Kodi uses for demuxing and decoding) logs tons of errors, it doesn't signal any error to the application level (Kodi). The only error Kodi sees in this case is audio stalling. Kodi tries a seek to re-sync but this file isn't even seekable using timestamps (I guess that the timestamps are not monotonic increasing), so playback jumps to the beginning.
> 
> > What I start to think is that, maybe the video player of kodi is born as file player so is more sensible to errors in streams than player like vlc that born as network player?
> 
> This applies to ffmpeg. In particular the mpeg-ts demuxer is not the best when it comes to streaming. In this case here a simple internal reset after the 100th error would most likely solve the issue but the demuxer doesn't try any error corrections. Nevertheless, ffmpeg implements the specifications and if some source does not play with ffmpeg, it's most likely a problem with the source.
> 
> > I understand, just wondering how vlc or other directshow based players manage this condition without issues.
> 
> Those directshow players do not use ffmpeg (vlc doesn't either). Windows world isn't famous for complying to specifications.
> Best solution for your problem is to file an issue at mediaportal

sorry, but i have to disagree with you

according with vlc logs it is a situation that can be managed at player level

>`main debug: VoutDisplayEvent 'resize' 1920x873
main debug: VoutDisplayEvent 'resize' 1920x873
direct3d11 debug: Detected size change 1552x873
main debug: resized to 1920x901
main debug: VoutDisplayEvent 'resize' 1920x901
main debug: VoutDisplayEvent 'resize' 1920x901
direct3d11 debug: Detected size change 1601x901
main debug: auto hiding mouse cursor
main debug: auto hiding mouse cursor
main debug: auto hiding mouse cursor
ts warning: discontinuity received 0x9 instead of 0x6 (pid=0)
ts error: libdvbpsi error (PSI decoder): TS discontinuity (received 9, expected 6) for PID 0
ts debug: PATCallBack called
ts debug: new PAT ts_id=4 version=2 current_next=1
ts debug: * number=137 pid=32
ts warning: discontinuity received 0x9 instead of 0x6 (pid=32)
ts error: libdvbpsi error (PSI decoder): TS discontinuity (received 9, expected 6) for PID 32
ts debug: PMTCallBack called for program 137
ts debug: new PMT program number=137 version=2 pid_pcr=48
ts debug: * pid=48 type=0x1b AVC video stream as defined in ITU-T Rec. H.264
ts debug: - ES descriptor tag 0x52
ts debug: => pid 48 has now es fcc=h264
ts debug: * pid=64 type=0x81 User Private
ts debug: - ES descriptor tag 0x52
ts debug: => pid 64 has now es fcc=a52 
main debug: killing decoder fourcc `mp4a'
main debug: removing module "faad"
mmdevice debug: state changed: 0
wasapi debug: reset
main debug: removing module "samplerate"
main debug: removing module "scaletempo"
main debug: removing module "float_mixer"
main debug: keeping audio output
main debug: removing module "mpeg4audio"
main debug: looking for packetizer module matching "any": 25 candidates
main debug: using packetizer module "a52"
main debug: looking for audio decoder module matching "any": 23 candidates
main debug: reusing audio output
main error: too low audio sample frequency (0)
main debug: keeping audio output
main error: failed to create audio output
avcodec debug: using ffmpeg Lavc58.6.103
avcodec debug: CPU flags: 0x000fd3db
avcodec debug: codec (ac3) started
main debug: using audio decoder module "avcodec"
ts debug: * pid=17 listening for SDT
ts debug: enabling pid 48 from program 137
ts debug: enabling pid 64 from program 137
ts debug: enabling pcr pid 48 from program 137
main error: ES_OUT_SET_(GROUP_)PCR is called too late (pts_delay increased to 1000 ms)
main debug: ES_OUT_RESET_PCR called
ts warning: discontinuity received 0x9 instead of 0x3 (pid=48)
main debug: Buffering 0%
main debug: Buffering 3%
main debug: Buffering 7%
main debug: Buffering 10%
main debug: Buffering 14%
main debug: Buffering 17%
avcodec debug: available hardware decoder output format 119 (cuda)
avcodec debug: available hardware decoder output format 53 (dxva2_vld)
main debug: Buffering 21%
avcodec debug: available hardware decoder output format 118 (d3d11va_vld)
avcodec debug: available hardware decoder output format 174 (d3d11)
avcodec debug: available software decoder output format 0 (yuv420p)
avcodec debug: reusing decoder output format 118
main debug: Buffering 24%
main debug: Buffering 28%
main debug: Buffering 31%
main debug: Buffering 35%
main debug: Buffering 38%
main debug: Buffering 42%
ts warning: discontinuity received 0x0 instead of 0x3 (pid=64)
main debug: Buffering 45%
a52 debug: A/52 channels:6 samplerate:48000 bitrate:384000
main debug: Buffering 49%
main debug: restarting module due to input format change
main debug: removing module "avcodec"
main debug: looking for audio decoder module matching "any": 23 candidates
main debug: reusing audio output
main error: module not functional
main debug: keeping audio output
main error: failed to create audio output
main debug: Buffering 53%
avcodec debug: using ffmpeg Lavc58.6.103
avcodec debug: CPU flags: 0x000fd3db
avcodec debug: codec (ac3) started
main debug: using audio decoder module "avcodec"
main debug: Buffering 56%
main debug: reusing audio output
main debug: looking for aout stream module matching "any": 2 candidates
main debug: Buffering 60%
main debug: Buffering 63%
main debug: Buffering 67%
main debug: Buffering 70%
main debug: Buffering 74%
main debug: Buffering 77%
main debug: Received first picture
main debug: Buffering 81%
main debug: Buffering 84%
main debug: Buffering 88%
main debug: Buffering 91%
main debug: Buffering 95%
main debug: Buffering 98%
main debug: Stream buffering done (1025 ms in 101 ms)
wasapi debug: buffer size : 96000 frames
wasapi debug: maximum latency: 000 ns
wasapi debug: default period : 10000000 ns
wasapi debug: minimum period : 3000000 ns
main debug: using aout stream module "wasapi"
main debug: output 'f32l' 48000 Hz 3F2M/LFE frame=1 samples/24 bytes
main debug: looking for audio volume module matching "any": 2 candidates
main debug: using audio volume module "float_mixer"
main debug: input 'f32l' 48000 Hz 3F2M/LFE frame=1 samples/24 bytes
main debug: looking for audio filter module matching "scaletempo": 16 candidates
scaletempo debug: format: 48000 rate, 6 nch, 4 bps, fl32
scaletempo debug: params: 30 stride, 0.200 overlap, 14 search
scaletempo debug: 1.000 scale, 1440.000 stride_in, 1440 stride_out, 1152 standing, 288 overlap, 672 search, 2400 queue, fl32 mode
main debug: using audio filter module "scaletempo"
main debug: conversion: 'f32l'->'f32l' 48000 Hz->48000 Hz 3F2M/LFE->3F2M/LFE
main debug: conversion pipeline complete
main debug: conversion: 'f32l'->'f32l' 48000 Hz->48000 Hz 3F2M/LFE->3F2M/LFE
main debug: conversion pipeline complete
main debug: looking for audio resampler module matching "any": 3 candidates
main debug: using audio resampler module "samplerate"
main debug: Decoder wait done in 281 ms
main debug: inserting 22822 zeroes
mmdevice debug: state changed: 1
main debug: incoming request - stopping current input`

it might be managed at server level (I tested also with npvr on server , same story even the ts files is created on the kodi client).

Maybe it's just matter of philosofy, but i think it's just a lack of recovering from ts discontinuities or errors of kodi video player.

Thank you anyway

**Comment by mrmojo666** [comment]
maybe I'm wrong again but seems that vlc is ffmpeg library based

[https://git.videolan.org/?p=vlc.git;a=blob;f=modules/codec/avcodec/fourcc.c](url)

**Comment by FernetMenta** [comment]
> maybe I'm wrong again but seems that vlc is ffmpeg library based

have a look at vlc's demuxer code, which is not ffmpeg:
https://git.videolan.org/?p=vlc.git;a=tree;f=modules/demux/mpeg;h=ef3d606dfb8055055ff5648684525f67ba476183;hb=HEAD

> Maybe it's just matter of philosofy, but i think it's just a lack of recovering from ts discontinuities or errors of kodi video player.

feel free to post your ideas of how this could be solved without doing harm to other scenarios. be aware that discontinuities in timestamps are a common scenario. same applies to appearing and disappearing streams.

**Comment by FernetMenta** [comment]
@mrmojo666 checkout the sample file provided here: https://github.com/xbmc/xbmc/issues/17080#issue-542749956

There is a change in audio format at around 2:00 min but the file is in spec. Kodi plays this file with a smooth transition, not loosing any audio.
vlc plays the same file with an annoying click sound at this position. That happens if you try to correct out-of-spec errors.

**Comment by mrmojo666** [comment]
Again, imho because all the others players are able to recover from ts files with errors inside. Only kodi and other ffmpeg based players can't. So even this situation might be out of "stardard" the results is that kodi can't be used as client for many tv servers that uses pvr hdmi cards (mediaportal, npvr etc).  Reading your previous post seems that vlc uses different demuxer, than kodi. The question is : kodi demuxer is kodi code or ffmpeg library code? . In case of ffmpeg demuxer code, i will go to open a issue on their project. At the other case we can look to vlc code to try to fix kodi demuxer.

**Comment by FernetMenta** [comment]
> The question is : kodi demuxer is kodi code or ffmpeg library code?

Kodi does not have its own demuxer. In your case it uses ffmpeg's mpeg-ts demuxer.

>  i will go to open a issue on their project

Try your luck. In general they don't implement workarounds for such kind of issues. But you might be able to convince them that making av_read_frame report an error to the application level would be a good idea.
